torch/utils/_pytree.py

```diff
@@ -0,0 +1,150 @@
+
+A *pytree* is Python nested data structure. It is a tree in the sense that
+nodes are Python collections (e.g., list, tuple, dict) and the leaves are
+Python values. Furthermore, a pytree should not contain reference cycles.
+
+pytrees are useful for working with nested collections of Tensors. For example,
+one can use `tree_map` to map a function over all Tensors inside some nested
+collection of Tensors and `tree_unflatten` to get a flat list of all Tensors
+inside some nested collection. pytrees are helpful for implementing nested
+collection support for PyTorch APIs.
+
+This pytree implementation is not very performant due to Python overhead
+To improve the performance we can move parts of the implementation to C++.
+"""
+
+# A NodeDef holds two callables:
+# - flatten_fn should take the collection and return a flat list of values.
+#   It can also return some context that is used in reconstructing the
+#   collection.
+# - unflatten_fn should take a flat list of values and some context
+#   (returned by flatten_fn). It returns the collection by reconstructing
+#   it from the list and the context.
+Context = Any
+pytree_t = Any
+flatten_fn_t = Callable[[pytree_t], Tuple[List, Context]]
+unflatten_fn_t = Callable[[List, Context], pytree_t]
+NodeDef = NamedTuple('NodeDef', [
+    ('flatten_fn', flatten_fn_t),
+    ('unflatten_fn', unflatten_fn_t),
+])
+
+SUPPORTED_NODES: Dict[Any, NodeDef] = {}
+
+def _register_pytree_node(typ: Any, flatten_fn: flatten_fn_t, unflatten_fn: unflatten_fn_t):
+    SUPPORTED_NODES[typ] = NodeDef(flatten_fn, unflatten_fn)
+
+def _dict_flatten(d):
+    return d.values(), list(d.keys())
+
+def _dict_unflatten(values, context):
+    return {key: value for key, value in zip(context, values)}
+
+def _list_flatten(d):
+    return d, None
+
+def _list_unflatten(values, context):
+    return list(values)
+
+def _tuple_flatten(d):
+    return list(d), None
+
+def _tuple_unflatten(values, context):
+    return tuple(values)
+
+_register_pytree_node(dict, _dict_flatten, _dict_unflatten)
+_register_pytree_node(list, _list_flatten, _list_unflatten)
+_register_pytree_node(tuple, _tuple_flatten, _tuple_unflatten)
+
+
+# A leaf is defined as anything that is not a Node.
+def _is_leaf(pytree: pytree_t) -> bool:
+    return type(pytree) not in SUPPORTED_NODES.keys()
+
+
+# A TreeSpec represents the structure of a pytree. It holds:
+# "type": the type of root Node of the pytree
+# context: some context that is useful in unflattening the pytree
+# children_specs: specs for each child of the root Node
+# num_leaves: the number of leaves
+class TreeSpec:
+    def __init__(self, typ: Any, context: Context, children_specs: List['TreeSpec']):
+        self.type = typ
+        self.context = context
+        self.children_specs = children_specs
+        self.num_leaves: int = sum([spec.num_leaves for spec in children_specs])
+
+    def __repr__(self):
+        return f'TreeSpec({self.type.__name__}, {self.context}, {self.children_specs})'
```

Comment:
Oh this is a good point; @zou3519 you should consider making this a dataclass, then you can print easily by converting the dataclass to a dictionary and then pretty-printing it.

Reply:
This is definitely not very readable for very nested TreeSpecs.

I'll handle this in a follow-up task (#46538), I'm not completely sure TreeSpec should be a dataclass (because there's some logic in `__init__` that tracks the number of leaves but that can just be moved outside of `__init__`)

Reply:
You can use `__post_init__` for that, it's pretty easy